target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts

```diff
@@ -0,0 +1,277 @@
+
+		reset {
+			label = "reset";
+			gpios = <&tlmm 4 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+
+		wps {
+			label = "wps";
+			gpios = <&tlmm 63 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		power: power {
+			label = "ens620ext:amber:power";
+			gpios = <&tlmm 58 GPIO_ACTIVE_LOW>;
+		};
+
+		zlan1 {
+			label = "ens620ext:green:zlan1";
+			gpios = <&tlmm 1 GPIO_ACTIVE_LOW>;
+		};
+
+		wlan2G {
+			label = "ens620ext:green:wlan2G";
+			gpios = <&tlmm 3 GPIO_ACTIVE_HIGH>;
+		};
+
+		wan5G {
+			label = "ens620ext:green:wlan5G";
+			gpios = <&tlmm 0 GPIO_ACTIVE_LOW>;
+		};
+
+		lan {
+			label = "ens620ext:green:lan";
+			gpios = <&tlmm 2 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&cryptobam {
+	status = "okay";
+};
+
+&blsp_dma {
+	status = "okay";
+};
+
+&tlmm {
+	serial_pins: serial_pinmux {
+		mux {
+			pins = "gpio60", "gpio61";
+			function = "blsp_uart0";
+			bias-disable;
+		};
+	};
+
+	spi_0_pins: spi_0_pinmux {
+		mux {
+			function = "blsp_spi0";
+			pins = "gpio55", "gpio56", "gpio57";
+			drive-strength = <12>;
+			bias-disable;
+		};
+
+		mux_cs {
+			function = "gpio";
+			pins = "gpio54";
+			drive-strength = <2>;
+			bias-disable;
+			output-high;
+		};
+	};
+};
+
+&blsp1_spi1 { /* BLSP1 QUP1 */
+	pinctrl-0 = <&spi_0_pins>;
+	pinctrl-names = "default";
+	status = "okay";
+	cs-gpios = <&tlmm 54 GPIO_ACTIVE_HIGH>;
+
+	flash@0 {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		/*
+		 * U-boot looks for "n25q128a11" node,
+		 * if we don't have it, it will spit out the following warning:
+		 * "ipq: fdt fixup unable to find compatible node".
+		 */
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		linux,modalias = "m25p80", "mx25l1606e", "n25q128a11";
+		spi-max-frequency = <24000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition0@0 {
+				label = "0:SBL1";
+				reg = <0x00000000 0x00040000>;
+				read-only;
+			};
+			partition1@40000 {
+				label = "0:MIBIB";
+				reg = <0x00040000 0x00020000>;
+				read-only;
+			};
+			partition2@60000 {
+				label = "0:QSEE";
+				reg = <0x00060000 0x00060000>;
+				read-only;
+			};
+			partition3@c0000 {
+				label = "0:CDT";
+				reg = <0x000c0000 0x00010000>;
+				read-only;
+			};
+			partition4@d0000 {
+				label = "0:DDRPARAMS";
+				reg = <0x000d0000 0x00010000>;
+				read-only;
+			};
+			partition5@e0000 {
+				label = "0:APPSBLENV"; /* uboot env*/
+				reg = <0x000e0000 0x00010000>;
+				read-only;
+			};
+			partition5@f0000 {
+				label = "0:APPSBL"; /* uboot */
+				reg = <0x000f0000 0x00090000>;
+				read-only;
+			};
+			partition5@180000 {
+				label = "0:ART";
+				reg = <0x00180000 0x00010000>;
+				read-only;
+			};
+			partition5@190000 {
+				label = "0:HLOS";
+				reg = <0x00190000 0x00500000>;
+			};
+			partition5@690000 {
+				label = "rootfs";
+				reg = <0x00690000 0x00FD0000>;
+			};
+			partition5@1660000 {
```

Comment:
I'm curious. Do you know what's in the failsafe partition? Is it another kernel+rootfs/initramfs or what does it do?

Reply:
Not sure. This is what smeminfo shows from uboot:
(IPQ40xx) # smeminfo
flash_type:             0x6
flash_index:            0x0
flash_chip_select:      0x0
flash_block_size:       0x10000
flash_density:          0x2000000
partition table offset  0x0
No.: Name             Attributes            Start             Size
  0: 0:SBL1           0x0000ffff              0x0          0x40000
  1: 0:MIBIB          0x002040ff          0x40000          0x20000
  2: 0:QSEE           0x0000ffff          0x60000          0x60000
  3: 0:CDT            0x0000ffff          0xc0000          0x10000
  4: 0:DDRPARAMS      0x0000ffff          0xd0000          0x10000
  5: 0:APPSBLENV      0x0000ffff          0xe0000          0x10000
  6: 0:APPSBL         0x0000ffff          0xf0000          0x90000
  7: 0:ART            0x0000ffff         0x180000          0x10000
  8: 0:HLOS           0x0000ffff         0x190000         0x500000
  9: rootfs           0x0000ffff         0x690000         0xfd0000
 10: failsafe         0x0000ffff        0x1660000         0x8f0000
 11: u-boot-env       0x0000ffff        0x1f50000          0x10000
 12: userconfig       0x0000ffff        0x1f60000          0xa0000

Reply:
Not much useful information there. Could you download the content of the failsafe partition to your PC and run "binwalk" on it? This should give us a overview of what's stored there.

Reply:
Tried a couple of different variants of binwalk.
Here's one version's output:
Binwalk v2.1.2-720f770
Craig Heffner, ReFirmLabs
https://github.com/ReFirmLabs/binwalk

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             device tree image (dtb)
228           0xE4            gzip compressed data, maximum compression, has original file name: "Image", from Unix, last modified: 2017-07-26 06:48:55
3252188       0x319FDC        device tree image (dtb)
3288960       0x322F80        device tree image (dtb)
3329280       0x32CD00        device tree image (dtb)
3369676       0x336ACC        device tree image (dtb)
3402876       0x33EC7C        device tree image (dtb)
3436220       0x346EBC        device tree image (dtb)
3469720       0x34F198        device tree image (dtb)